core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -221,18 +220,16 @@ public long getLastComputedLeftDelayNanos() {
 
     /**
      * Updates delay left based on current time (in nanoseconds) and index/node settings.
-     * Should only be called from ReplicaShardAllocator.
      * @return updated delay in nanoseconds
      */
     public long updateDelay(long nanoTimeNow, Settings settings, Settings indexSettings) {
-        long delayTimeoutMillis = getAllocationDelayTimeoutSetting(settings, indexSettings);
+        long delayTimeoutNanos = getAllocationDelayTimeoutSettingNanos(settings, indexSettings);
         final long newComputedLeftDelayNanos;
-        if (delayTimeoutMillis == 0l) {
+        if (delayTimeoutNanos == 0l) {
             newComputedLeftDelayNanos = 0l;
         } else {
             assert nanoTimeNow >= unassignedTimeNanos;
-            long delayTimeoutNanos = TimeUnit.NANOSECONDS.convert(delayTimeoutMillis, TimeUnit.MILLISECONDS);
-            newComputedLeftDelayNanos = Math.max(0l, delayTimeoutNanos - (nanoTimeNow - unassignedTimeNanos));
+            newComputedLeftDelayNanos = delayTimeoutNanos - (nanoTimeNow - unassignedTimeNanos);
```

Comment:
Sorry, I don’t follow. When can this be negative? (we assert nanoTime >= unassignedTimeNanos)

> On 20 Nov 2015, at 16:24, Yannick Welsch <notifications@github.com> wrote:
> 
> In core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java:
> 
> >              newComputedLeftDelayNanos = 0l;
> >          } else {
> >              assert nanoTimeNow >= unassignedTimeNanos;
> > -            long delayTimeoutNanos = TimeUnit.NANOSECONDS.convert(delayTimeoutMillis, TimeUnit.MILLISECONDS);
> > -            newComputedLeftDelayNanos = Math.max(0l, delayTimeoutNanos - (nanoTimeNow - unassignedTimeNanos));
> > +            newComputedLeftDelayNanos = delayTimeoutNanos - (nanoTimeNow - unassignedTimeNanos);
> 
> This can be negative. This breaks consistency as the documentation for lastComputedLeftDelayNanos says // how long to delay shard allocation, not serialized (always positive, 0 means no delay).
> 
> —
> Reply to this email directly or view it on GitHub.
>